edb/server/connpool/pool.py

```diff
@@ -841,6 +842,27 @@ def _tick(self) -> None:
                     self._log_to_snapshot(
                         dbname=block.dbname, event='reset-quota')
 
+            if not was_starving and self._new_blocks_waitlist:
+                # Mode D assumes all connections are already in use or to be
+                # used, depending on their `release()` to schedule transfers.
+                # When just entering Mode D, there can be a special case when
+                # no further `release()` will be called because all acquired
+                # connections were returned to the pool before `_tick()` got a
+                # chance to set `self._is_starving`, while some other blocks
+                # are literally starving to death (blocked forever).
+                #
+                # This branch handles this particular case, by stealing
+                # connections from the idle blocks and try to free them into
+                # the starving blocks.
+
+                for block in list(self._blocks.values()):
+                    nconns = block.count_conns()
+                    num_waiters = block.count_waiters()
+                    if nconns > num_waiters:
+                        if (conn := block.try_steal()) is not None:
```

Comment:
Any particular reason we steal just one per block?

Reply:
good point, it could steal at most `nconns - num_waiters` times